File: Python3/667. Beautiful Arrangement II.py
class Solution:
    def constructArray(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[int]
        """
        # ans = list(range(1, n+1)) # 828ms 效率极低
        # for i in range(1, k):
        #     ans = ans[: i] + ans[i:][::-1]
        # return ans
        
        ans = list(range(1, n-k+1)) # 68ms
        i, j = n, n-k+1
        while i >= j:
            if i >= j:
                ans.append(i)
                i -= 1
            if i >= j:
                ans.append(j)
                j += 1
        return ans

File: Golang/667. Beautiful Arrangement II.go
func constructArray(n int, k int) []int {
    if k == 2 {
        ans := make([]int, n)
        ans[0], ans[1], ans[2] = 1, 3, 2
        for i := 3; i < n; i++ {
            ans[i] = i + 1
        }
        return ans
    }
    
    i := 1
    j := n
    ans := make([]int, n)
    count := 0
    reverse := false
    for k > 0 {
        if k > 0 {
            ans[count] = i
            count++
            i++
            k--
            reverse = false
        }
        if k > 0 {
            ans[count] = j
            count++
            j--
            k--
            reverse = true
        }
    }
    if reverse {
        for num := j; num >= i; num-- {
            ans[count] = num
            count++
        }
    } else {
        for num := i; num <= j; num++ {
            ans[count] = num
            count++
        }
    }
    return ans
}
